### ISSUE TITLE
Active malicious PasteJacking exploits in-the-wild affecting user security owing to lack of sufficient consideration to identified and other security concerns

### ISSUE THREAD

**Issue by gitcnd** [w3c/clipboard-apis#142]
The use-case for permitting a browser to store different content to that which a user expects or intends is strenuous at best, and every case identified is easily implemented with alternative mechanisms.

Additionally, it is debatable whether it is *ever* possible for a paste buffer to recieve something different to that which a user copied in a way that any user might consider "what they intended".  If you include the fact that no other applications in existence ever do this already, the debate looks unwinnable: no user would ever expect "copy/paste" in a browser to behave differently to ever other time they've ever use copy/paste before.

Multiple sites are now injecting malicious content into pasted operations - including modified-source-code websites (inserting maliciously altered code) and modified-CLI-sites (inserting linux command line operations that a user did not want).  Many other security considerations are also missing for your examples - like alteration of payment address (account numbers, cryptocurrency addresses, financial numbers like amounts (e.g. "$12.37" .vs. "$1237")), CR/TAB characters which can cause actions to occur immediately, subtle alteration of computer code to perform malicous actions (install malware, steal money, siphone off cryptocurrencies), and many more.

Section 10.1. "Pasting HTML and multi-part data" fail to even consider that *other* applications besides browsers will be consuming this potentially malicious content (despite that being one of the named use cases).

There are no identified use cases that warrant the introduction of the malicious opportunities that PasteJacking permits:

* Metadata - not need - "the source of the copied content" can be shown in text on the content being copied - if the user wants that, they can explicitly select and copy it.

* Rich content editing - not needed - "text which contains hyperlinks or other structure" is *already preserved* when pasted into applications that understand how to preserve it.  No website could ever know where a user intends to paste copied content, so there is no possible use case for that site to "reformat the content to preserve important information", because the site doesn't know the capabilities of where the content will be pasted into.  In most situations, even attempting to "preserve important information" is more than likely going to damage that information (if the target system understands how to accept a rich paste, having that richness removed damages the very information this proposal is supposed to support).  Users who want a rich paste, are going to paste into an application that accepts it.

* Graphics with built-in semantics - not needed - same as above - the website can never know where pasted text is intended for.  Math sites which wish to allow users "copy/paste" features for other applications can easily implement a suitable mechanism by giving the user a choice of what kind of target their copy operation is intended for.  There is no use case for doing this automatically, because the web page can never automatically know the capabilities of the intended paste target.

"the appropriate transformation occurring at paste time." - this is impossible, because no web site can know the paste-target capabilities without first asking the user (which is easy to do, and does not require exposing users to PasteJacking attacks for webmasters to implement).

Consider "\n rm -rf / \n" being added to a linux command line instructable !

In addition - "copy" is not the only possible action.  Sometimes "cut" takes place.  Alterations to what was "cut" in this situation results in "important information" becoming irrevocably lost - that is directly the opposite result to one of the intended use cases!

I propose:

Remove all capability for browsers to insert content into clipboards that has not been explicitly chosen by the user to be copied (and/or cut).  No modification. No addition.  Receiving something different (and potentially malicious, and almost always unwanted) into a paste buffer to that which was selected by the user is near-universally unintended, universally unexpected, potentially catastrophic, and does not have any demonstrated use-case that is not already easily implemented using alternative and safe existing mechanisms.

**Comment by gitcnd** [comment]
I thought of a compromise, in the (unlikely?) event that anyone actually might still want paste-buffer smashing:  relegate the altered version into a NON DEFAULT option of its own (e.g. right-click, and "copy" - works as expected - no alterations/addition, or right-click and "copy-suggested" (or some similar name) so that if a user actually wants to get something different to what they marked, they still have the option to explicitly request that (and therefore will be aware it took place - and the risks involved).

It might also be a good idea to warn them when they use this feature, such as:
a) pop up a message like "beware that malicious content might have been copied to your clipboard.  Take precautions before pasting this content"
b) show an interim window with a similar warning above, showing the user what has actually been copied

**Comment by othermaciej** [comment]
> Remove all capability for browsers to insert content into clipboards that has not been explicitly chosen by the user to be copied (and/or cut). No modification. No addition.

This wouldn't make a material difference to ability to put something unexpected on the pasteboard. There's plenty of ways to make HTML content look like one piece of text visually, but actually have different text contents when converted to plaintext. (e.g. invisible text on top of a background with different text or an image of text). It seems like the most dangerous contexts for paste alteration are plaintext contexts.

**Comment by hallvors** [comment]
No-longer-editing-this-spec drive-by comment: I'm sure suggestions for improving the spec's prose regarding security considerations can be useful. It may be worth it to say explicitly what @othermaciej laid out above: it is generally held to be true that sites can already control plain text and rich text content on the clipboard through manipulating the selection without this API, for example on keydown/mousedown/contextmenu events that might indicate a copy event. Hence the API gives "good" sites a cleaner way to do what they might need without actually granting "bad" sites meaningful extra powers. Users who are technical enough to paste stuff into a command line are - unfortunately - hard to protect if they are not also skilled enough to understand the risks..

Warnings if the site modified what goes on the clipboard is likely an excellent idea (though if many sites do it for good purposes, warning fatigue will occur). But this is user-agent UI-land, not spec material. User-agents are supposed to innovate and differentiate themselves on UI and security mechanisms. One might choose to block all clipboard API usage unless a site gets white-listed, another might choose to show a popup "site changed clipboard, do you want to copy selected content instead?". Then the market can decide what security features win. Most of the lower-level technical specs try to not impose UI requirements because of an overall idea that this is a good separation of concerns. 

In the security field, it is indeed less easy to separate concerns in the first place.. lots of security threats are due to poor UI or need mitigation with UI. But that's the historical background.

**Comment by gitcnd** [comment]
@sspi - I can't see how that could ever work - there's zero chance you could ever update every-application-ever-written-that-accepts-paste-input such that "It is then the responsibility of the ... application reading the clipboard to ensure with the user that the pasted content is correct."

And that's not even *starting* on how it might even be possible to DO that in the first place - consider some copy/pasted code that uses variable names containing l or I or other nearly/completely indistinguishable characters...

@othermaciej - you can't overlay text I want to select with an image, because then the text can't be marked for copy.  You can't use "invisible text" because it shows up when being marked.  It is trivially easy to overcome other "tricks" that might exist to accomplish PasteJacking via other means (for example: clearing a paste-selection should anything modify what was selected - making it impossible to select the wrong thing).

For the record - discussing reasons not to remove this unnecessary security problem makes no sense without full consideration of the use case in the first instance.  I already demonstrated that no legitimate use case has been suggested which doesn't already have a more-desirable workaround already.

Dangerous features should be removed when there is no need for their existence.  Even if other similar dangers still exist, that is no excuse for keeping the unwanted dangerous feature in the first place (but, bonus, if such other dangers DO still exist, now would be a great time to address and remove those security issues at the same time).

**Comment by othermaciej** [comment]
@githcnd here's a POC  for pure CSS pastejacking that works in Mac Safari, Firefox, and Chrome: https://jsfiddle.net/7gp6yu34/

**Comment by gitcnd** [comment]
Nice work - not fully functional on Chrome, but close - shows both text over top of one-another after the copy:
https://mega.nz/file/8xNXXIIQ#NDv_YexuOrgAu6PjL96HAiotP8WboqloaBkyVEIn3AE
as well as the non-overlapped text during the mark

Like I said before though - finding some other exploit in a browser does not excuse anyone from allowing one that is currently being exploited to remain - most especially not when there is no valid use case for the PasteJacking "feature" that should be removed in the first place.

**Comment by hallvors** [comment]
Anyone copying from, say, Google Docs with a CANVAS-based rendering engine depends on these features. The use cases are plenty, and unless you have specific and useful suggestions to improving the "security consideration" prose in the spec, I recommend the current editors close this issue.

You may want to ask your favourite browser to implement a "disable clipboard API" setting or write a browser extension that stops websites from using it. I think both are excellent ideas for the cautious. However, arguing that the spec should simply drop widely requested and widely used functionality is, respectfully, wasted time.
